### PR TITLE
Feature/automate api tests

### PR DESCRIPTION
**Pull request by mdp0**
## Description
Add Mocha API test run against docker compose environment

### Changes
- [ARTS-408] - Introduce Mocha tests. Add ci target for Mocha.

### Checklist:

- [ ] Branch name follows convention (feature/arts-###-short-name OR fix/arts-###-short-name)
- [ ] I have included a short-meaningful title, description and changes for this PR

[ARTS-408]: https://ndph-arts.atlassian.net/browse/ARTS-408

### CODE REVIEW

**Comment by mdp0** [comment]
WiP changes. Need to add publish to GHCR back in.

File: api-tests/package.json [hunk 1]
```diff
@@ -7,7 +7,7 @@
     "test:awesome": "mocha specs/*/*.js --reporter mochawesome || true",
     "open:report": "open mochawesome-report/mochawesome.html",
     "test": "mocha specs/*/*.js",
-    "test-ci": "mocha specs/*/*.js --reporter mocha-junit-reporter"
+    "api-test-ci": "npm i && mocha --timeout 8000 --reporter mocha-junit-reporter --reporter-options mochaFile=./api-test-results.xml specs/practitioner-service/*.js || true"
```
Comment (shajidissa) [inline api-tests/package.json:10]: practitioner-service should be replace by *

to include other tests ?

File: api-tests/specs/practitioner-service/createperson.js [hunk 2]
```diff
@@ -2,15 +2,16 @@ const requests = require('../../data/practitioner-service/createperson')
...
+    baseRequest = request(conf.baseUrl)
 })
 
 describe('As a user with Create Person permission, I want to have my create person request validated by the system, So that I cannot create an invalid person', function () {
 
     it('When I submit an API request to create a Person with a value for all mandatory fields and no fields exceed their specified maximum length, Then a new Person record is persisted in the system with a unique identifier And I receive a success acknowledgement', async () => {
         const response = await baseRequest.post('/practitioner').send(requests.validPerson);
         expect(response.text).to.contain("id")
-        expect(response.status).to.equal(HttpStatus.CREATED)
+        expect(response.status).to.equal(201)
```
Comment (shajidissa) [inline api-tests/specs/practitioner-service/createperson.js:14]: better to leave it as HttpStatus.CREATED ?

**Review by shajidissa — CHANGES_REQUESTED** [review]

**Review by shajidissa — CHANGES_REQUESTED** [review]
Mark theres some white space changes in the roles and trial-config-service tests.. I'm sure you didn't mean to touch these files.

File: api-tests/package.json (hunk 1 above)
Comment (mdp0) [reply api-tests/package.json:10]: Had no role service to try that against, it's in compose now - great catch, I'll add it in

File: .github/workflows/docker-build-push.yml [hunk 3]
```diff
@@ -71,15 +60,25 @@ jobs:
 
   test:
```
Comment (tamirkamara) [inline .github/workflows/docker-build-push.yml:61]: Can you please add an "if" statement on this job to only run on PR events (we don't need to it run on the nightly release for example)?
You can see an example on this same file once it's updated with the version currently on the main branch.

File: api-tests/data/common/headers.js [hunk 4]
```diff
@@ -1,4 +0,0 @@
-module.exports.jsonContent = {
-    'Content-Type': 'application/json',
-    'X-Version': 1.0,
-}
```
Comment (sameerapurini) [inline api-tests/data/common/headers.js:4]: Hi @mdp0 not sure why this file was deleted?? I have added this in order to get the header components in the request to be read by the other services.

I might need a walkthrough of the PR as many changes have been done on the api tests that qa have written.

**Review by sameerapurini — COMMENTED** [review]
Hi @mdp0 the http statuses and the other assertions in the test files have been taken into consideration after discussing with the relevant developer @KateSant and @balteravishay , can you please revert those changes, thanks

And also the tests are not running after your changes. I have updated the readme as per the way I set up the api tests, please revert the changes in those files as well.

I will go through them with the other QA and yourself to update them.

these are the files to be reverted
1. practitioner-service.js
2. roleservice.js
3. put back the deleted header, js
4. The mochawesome report path in the package.json file

File: api-tests/package.json [hunk 5]
```diff
@@ -4,10 +4,10 @@
   "description": "A base repo for all microservices",
   "main": "index.js",
   "scripts": {
-    "test:awesome": "mocha specs/*/*.js --reporter mochawesome || true",
```
Comment (sameerapurini) [inline api-tests/package.json:7]: updating this has broken the tests, please revert it.

File: api-tests/specs/role-service/roleservice.js [hunk 6]
```diff
@@ -1,39 +1,36 @@
...
     it('User is shown a bad request error when the role name field exceeds the expected string length', async () => {
-        const response = await baseRequest.post('/roles').send(requests.tooLongId);
+        const response = await baseRequest.post(endpointUri).send(requests.tooLongId);
         expect(response.status).to.equal(HttpStatus.BAD_REQUEST);
         expect(response.text).to.contain("Role ID is too long")
     });
 
     it('User is able to view the created roles', async () => {
-        const response = await baseRequest.get('/roles?page=0&size=2');
+        const response = await baseRequest.get(`${endpointUri}?page=0&size=2`);
```
Comment (sameerapurini) [inline api-tests/specs/role-service/roleservice.js:33]: We usually have the end point url clearly in the request in order to differentiate between the different services. I would recommend to revert it.

**Review by sameerapurini — CHANGES_REQUESTED** [review]

File: api-tests/package.json (hunk 5 above)
Comment (mdp0) [reply api-tests/package.json:7]: Restored accidentally deleted test path!

File: api-tests/specs/role-service/roleservice.js (hunk 6 above)
Comment (mdp0) [reply api-tests/specs/role-service/roleservice.js:33]: Agreed we need to specialize in each test, until we have a single gateway

File: api-tests/specs/practitioner-service/createperson.js (hunk 2 above)
Comment (mdp0) [reply api-tests/specs/practitioner-service/createperson.js]: Definitely. Updated

File: api-tests/data/common/headers.js (hunk 4 above)
Comment (mdp0) [reply api-tests/data/common/headers.js:4]: Removed as it wasn't in use anywhere.

Comment (sameerapurini) [reply api-tests/data/common/headers.js:4]: I was using it to export the json header but right now exporting the modules seems to be doing the job :)

**Review by tamirkamara — APPROVED** [review]

**Review by sameerapurini — APPROVED** [review]
I will update the readme file in a different pr

**Pull request closed without merging**